### PR TITLE
fix: coalesce should return correct datatype

### LINKED ISSUE #167: `coalesce` function returns incorrect result
### Describe the bug

When I work on SortMergeJoin, there are some TPCDS query failures happened with errors like:

```
- q38 *** FAILED ***
  java.lang.Exception: Expected "struct<[count(1):bigint]>", but got "struct<[]>" Schema did not match
SELECT count(*)
FROM (
       SELECT DISTINCT
         c_last_name,
         c_first_name,
         d_date
       FROM store_sales, date_dim, customer
       WHERE store_sales.ss_sold_date_sk = date_dim.d_date_sk
         AND store_sales.ss_customer_sk = customer.c_customer_sk
         AND d_month_seq BETWEEN 1200 AND 1200 + 11
       INTERSECT
       SELECT DISTINCT
         c_last_name,
         c_first_name,
         d_date
       FROM catalog_sales, date_dim, customer
       WHERE catalog_sales.cs_sold_date_sk = date_dim.d_date_sk
         AND catalog_sales.cs_bill_customer_sk = customer.c_customer_sk
         AND d_month_seq BETWEEN 1200 AND 1200 + 11
       INTERSECT
       SELECT DISTINCT
         c_last_name,
         c_first_name,
         d_date
       FROM web_sales, date_dim, customer
       WHERE web_sales.ws_sold_date_sk = date_dim.d_date_sk
         AND web_sales.ws_bill_customer_sk = customer.c_customer_sk
         AND d_month_seq BETWEEN 1200 AND 1200 + 11
     ) hot_cust
LIMIT 100
...
org.apache.comet.CometNativeException
Arrow error: Invalid argument error: RowConverter column schema mismatch, expected Utf8 got Date32
```

It is because DataFusion `coalesce` function returns a `Date32` array from `Date32` inputs (this is correct) but its return type is `Utf8`. The details are in https://github.com/apache/arrow-datafusion/issues/9458. The fix is at https://github.com/apache/arrow-datafusion/pull/9459.

### Steps to reproduce

```rust
test("coalesce should return correct datatype") {
     Seq(true, false).foreach { dictionaryEnabled =>
       withTempDir { dir =>
         val path = new Path(dir.toURI.toString, "test.parquet")
         makeParquetFileAllTypes(path, dictionaryEnabled = dictionaryEnabled, 10000)
         withParquetTable(path.toString, "tbl") {
           checkSparkAnswerAndOperator(
             "SELECT coalesce(cast(_18 as date), cast(_19 as date), _20) FROM tbl")
         }
       }
     }
   }
```

### Expected behavior

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by viirya**
## Which issue does this PR close?

Closes #167.

### CODE REVIEW

File: spark/src/test/scala/org/apache/comet/CometExpressionSuite.scala [hunk 1]
```diff
@@ -34,6 +34,19 @@ import org.apache.comet.CometSparkSessionExtensions.{isSpark32, isSpark33Plus, i
...
+      withTempDir { dir =>
+        val path = new Path(dir.toURI.toString, "test.parquet")
+        makeParquetFileAllTypes(path, dictionaryEnabled = dictionaryEnabled, 10000)
+        withParquetTable(path.toString, "tbl") {
+          checkSparkAnswerAndOperator(
+            "SELECT coalesce(cast(_18 as date), cast(_19 as date), _20) FROM tbl")
+        }
+      }
+    }
+  }
```
Comment (viirya) [inline spark/src/test/scala/org/apache/comet/CometExpressionSuite.scala:48]: Due to the issue https://github.com/apache/arrow-datafusion/issues/9458, the return type and the actual output array is different in DataFusion `coalesce` function:

```
  org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 2.0 failed 1 times, most recent failure: Lost task 0.0 in stage 2.0 (TID 2) (192.168.86.44 executor driver): org.apache.comet.CometNativeException
: Arrow error: Invalid argument error: column types must match schema types, expected Utf8 but found Date32 at column index 0                                                                                                          
        at org.apache.comet.Native.executePlan(Native Method)                                                      
        at org.apache.comet.CometExecIterator.executeNative(CometExecIterator.scala:65)
        at org.apache.comet.CometExecIterator.getNextBatch(CometExecIterator.scala:111)
        at org.apache.comet.CometExecIterator.hasNext(CometExecIterator.scala:126)
        at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage1.columnartorow_nextBatch_0$(Unknown Source)
        at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage1.processNext(Unknown Source)
        at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
        at org.apache.spark.sql.execution.WholeStageCodegenExec$$anon$1.hasNext(WholeStageCodegenExec.scala:760)
```

**Comment by viirya** [comment]
cc @sunchao

**Review by sunchao — APPROVED** [review]
LGTM

**Comment by viirya** [comment]
Merged. Thanks.

**Pull request merged**